### PR TITLE
Feature/517 glossary cron job

### PR DESCRIPTION
**Pull request by maxdiebold-erg**
## Related Issues:
* [HMW-517](https://jira.epa.gov/browse/HMW-517)

## Main Changes:
* Created a `cron` task (`node-cron`) that fetches and caches glossary content to the public S3 bucket daily. When not running locally, this task is assigned to the first server instance.
* Updated the client-side glossary components to pull glossary terms from the cached data, rather than the terminology service.

## Steps To Test:
1. Start the application, and verify no error messages are printed to the terminal console.
2. Verify the glossary panel and terms work.
3. In `server.js`, change **line 161** to `'* * * * *',`.
4. Restart the application, wait one minute, then confirm that "Running glossary cron task" is logged.
5. Stop the application.
6. Run the command `npm run glossary` in the app or server directories. Confirm the glossary task completes successfully.

## TODO:
- [ ] Test the S3 configuration after merging.
- [ ] Test that the cron job is only scheduled for one instance after merging.
- [ ] Restore the number of instances in `manifest-dev.yml` after testing the previous two items.

### CODE REVIEW

**Comment by maxdiebold-erg** [comment]
Looks like there's an issue with the glossary terms on the Data page not triggering a fetch, I'll need to look into it.

**Comment by cschwinderg** [comment]
Just a heads up. The map popups have a similar issue. There it is an infinite loading spinner.

**Comment by maxdiebold-erg** [comment]
@cschwinderg Good catch, it was a context issue after all.

**Review by cschwinderg — CHANGES_REQUESTED** [review]
Works well locally! It will be awesome to move away from Synaptica somewhat. I haven't looked very closely at the front-end side, since you are still working on that. The below comments are all focused on the server side. 

I pointed my local environment to our dev S3 bucket and found some issues, I left some comments below on how to fix that. I was considering just committing my code, but I ended up not doing that on the off chance that you are updating those files.

Here are some other suggestions:
* run prettier on the following files: `manifest-staging.yml`, `manifest-attains.yml`, `manifest-production.yml`.
* In the `services.json` and `services-attains.json` files, change the `glossaryURL` value to `https://etss.epa.gov/synaptica_rest_services/api/vocabs/name/HMW%20Glossary/terms/full`. This gets us away from the saved report. 
  * Currently, the saved report only runs once every 24 hours. So if Kiki, adds in new glossary terms we would have to wait up to 24 hours to update the glossary. Switching to the plain web service call, allows us to update the glossary immediately after Kiki adds in new terms.

File: app/server/app/server/utilities/s3.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+      // write the file
+      writeFileSync(`${subFolderPath}/${filePath}`, fileToUpload);
+    } else {
+      // setup public s3 bucket
+      const s3 = getS3Client();
+
+      // upload the file
+      const command = new PutObjectCommand({
+        Bucket: getS3Config().bucket,
+        Key: `${subFolder}/${filePath}`,
```
[the remark this answers is not in the document]
Comment (maxdiebold-erg) [reply app/server/app/server/utilities/s3.js]: Good point! I'll go ahead and use the suggestion, so we can keep the same structure.

**Comment by cschwinderg** [comment]
On the client side. It looks like the glossary items in the map popup have been fixed. Now I'm seeing some new issues elsewhere.

* Data page - If you navigate directly to the data page, the glossary terms load, but clicking on them doesn't filter to the clicked term
  * If you open the data page from the community page, then everything works fine.
* About page - If you navigate directly to the about page, the glossary terms have infinite loading spinners.
  * If you open the data page from the community page, then everything works fine.
* Waterbody report page - Infinite loading spinners for some terms and the term in the `Waterbody Condition` item doesn't filter the glossary.
* Monitoring report page - Terms don't filter the glossary.
* Plan Summary - Terms don't filter the glossary.

**Comment by maxdiebold-erg** [comment]
I've been on this new kick after reading https://react.dev/learn/you-might-not-need-an-effect, but it's less intuitive than I thought.

**Comment by cschwinderg** [comment]
I think we did go a little overboard with effects in some places. 

The worst offender is probably the water quality overview tab on the state & tribal page. I think moving most of that code out of effects, would probably fix the jitteryness with the charts on that page and probably simplify the code somewhat.

**Review by cschwinderg — APPROVED** [review]
Seems like you fixed the issues!

**Comment by cschwinderg** [comment]
Actually, I do see an issue. If you block the service call for getting the glossary data, then the terms show infinite loading spinners and the glossary button in the banner stays greyed out.

On production all of the glossary stuff works like normal, but when the glossary opens we display an error in the glossary. I think we want to keep this behavior.

**Comment by maxdiebold-erg** [comment]
Good call, I reverted to the previous behavior. Seems a bit roundabout, but if it ain't broke...

**Pull request merged**